File: tif_customization/tif_customization/page/field_staff_visit_mo/field_staff_visit_mo.py
import json

import frappe
from frappe import _
from frappe.utils import getdate


@frappe.whitelist()
def get_report_data(filters=None):
	"""Return Field Staff Visit report data grouped by section and user."""
	try:
		if isinstance(filters, str):
			filters = json.loads(filters)
		elif not filters:
			filters = {}

		from_date, to_date = _parse_dates(filters)
		user = (filters.get("user") or "").strip() or None
		section = (filters.get("section") or "").strip() or None

		marketing = _get_marketing_visits(from_date, to_date, user=user, section=section)
		me_visits = _get_me_visits(from_date, to_date, user=user, section=section)
		training = _get_training_sessions(from_date, to_date, user=user, section=section)

		return {
			"from_date": str(from_date),
			"to_date": str(to_date),
			"user": user,
			"section": section,
			"marketing": marketing,
			"me": me_visits,
			"training": training,
		}
	except Exception as e:
		frappe.log_error(f"Error in get_report_data: {str(e)}", "Field Staff Visit Report")
		return {"error": str(e)}


def _parse_dates(filters):
	from_date = filters.get("from_date")
	to_date = filters.get("to_date")

	if from_date:
		from_date = getdate(from_date)
	if to_date:
		to_date = getdate(to_date)

	if not from_date or not to_date:
		# Backward compatibility with month/year filters
		month = int(filters.get("month") or 0) or getdate().month
		year = int(filters.get("year") or 0) or getdate().year
		from frappe.utils import get_first_day, get_last_day

		from_date = get_first_day(f"{year}-{month:02d}-01")
		to_date = get_last_day(from_date)

	if from_date > to_date:
		frappe.throw(_("From Date cannot be after To Date."))

	return from_date, to_date


def _user_names(users):
	users = sorted({u for u in users if u})
	if not users:
		return {}
	rows = frappe.get_all("User", filters={"name": ["in", users]}, fields=["name", "full_name"])
	return {row.name: (row.full_name or row.name) for row in rows}


def _section_user_filters(user=None, section=None):
	conditions = []
	params = {}
	if user:
		conditions.append(
			"""(
				TRIM({user_expr}) = %(user)s
				OR u.name = %(user)s
				OR LOWER(TRIM(u.full_name)) = LOWER(%(user)s)
			)"""
		)
		params["user"] = user
	if section:
		conditions.append("COALESCE(e.department, 'Unassigned') = %(section)s")
		params["section"] = section
	return conditions, params


def _aggregate_section_user_rows(rows, metrics_fn):
	"""Group SQL rows by section + user and apply metrics_fn(bucket, row)."""
	buckets = {}
	for row in rows:
		section = row.get("section") or "Unassigned"
		user = row.get("user") or "Unassigned"
		key = (section, user)
		bucket = buckets.setdefault(
			key,
			{
				"section": section,
				"user": user,
				"user_name": user,
			},
		)
		metrics_fn(bucket, row)

	users = {bucket["user"] for bucket in buckets.values() if bucket["user"] not in ("Unassigned", "")}
	name_map = _user_names(users)
	for bucket in buckets.values():
		if bucket["user"] in name_map:
			bucket["user_name"] = name_map[bucket["user"]]

	result_rows = sorted(
		buckets.values(),
		key=lambda item: (item["section"].lower(), item["user_name"].lower()),
	)
	return result_rows


def _employee_join(user_expr):
	return f"""
		LEFT JOIN `tabUser` u ON (
			u.name = TRIM({user_expr})
			OR LOWER(TRIM(u.full_name)) = LOWER(TRIM({user_expr}))
		)
		LEFT JOIN `tabEmployee` e ON (
			e.user_id = u.name
			OR LOWER(TRIM(e.employee_name)) = LOWER(TRIM({user_expr}))
		)
	"""


def _get_marketing_visits(from_date, to_date, user=None, section=None):
	user_expr = "COALESCE(NULLIF(TRIM(fv.visit_by), ''), fv.owner)"
	extra_conditions, extra_params = _section_user_filters(user=user, section=section)
	extra_sql = ""
	if extra_conditions:
		extra_sql = " AND " + " AND ".join(c.replace("{user_expr}", user_expr) for c in extra_conditions)

	rows = frappe.db.sql(
		f"""
		SELECT
			COALESCE(e.department, 'Unassigned') AS section,
			{user_expr} AS user,
			COALESCE(fv.marketing_visit_category, 'Unspecified') AS category,
			COUNT(*) AS cnt
		FROM `tabField Visit` fv
		{_employee_join(user_expr)}
		WHERE fv.docstatus < 2
		AND fv.type = 'Marketing'
		AND COALESCE(fv.visit_date, DATE(fv.timestamp)) BETWEEN %(from_date)s AND %(to_date)s
		{extra_sql}
		GROUP BY COALESCE(e.department, 'Unassigned'), {user_expr}, COALESCE(fv.marketing_visit_category, 'Unspecified')
		""",
		{
			"from_date": from_date,
			"to_date": to_date,
			**extra_params,
		},
		as_dict=True,
	)

	def apply_metrics(bucket, row):
		bucket.setdefault("new", 0)
		bucket.setdefault("followup", 0)
		bucket.setdefault("tps", 0)
		category = row.get("category")
		cnt = row.get("cnt", 0)
		if category == "New":
			bucket["new"] += cnt
		elif category == "Followup & Other Visits":
			bucket["followup"] += cnt
		elif category == "TPS Visits":
			bucket["tps"] += cnt

	result_rows = _aggregate_section_user_rows(rows, apply_metrics)
	totals = {"new": 0, "followup": 0, "tps": 0, "total": 0}
	for row in result_rows:
		row["total"] = row.get("new", 0) + row.get("followup", 0) + row.get("tps", 0)
		totals["new"] += row.get("new", 0)
		totals["followup"] += row.get("followup", 0)
		totals["tps"] += row.get("tps", 0)
		totals["total"] += row.get("total", 0)

	return {"rows": result_rows, "totals": totals}


def _get_me_visits(from_date, to_date, user=None, section=None):
	user_expr = "COALESCE(NULLIF(TRIM(fv.me_visit_by), ''), fv.owner)"
	extra_conditions, extra_params = _section_user_filters(user=user, section=section)
	extra_sql = ""
	if extra_conditions:
		extra_sql = " AND " + " AND ".join(c.replace("{user_expr}", user_expr) for c in extra_conditions)

	rows = frappe.db.sql(
		f"""
		SELECT
			COALESCE(e.department, 'Unassigned') AS section,
			{user_expr} AS user,
			COALESCE(fv.me_activity_status, 'Unspecified') AS status,
			COUNT(*) AS cnt
		FROM `tabField Visit` fv
		{_employee_join(user_expr)}
		WHERE fv.docstatus < 2
		AND fv.type = 'M&E'
		AND COALESCE(fv.me_visit_date, fv.me_starting_date, DATE(fv.me_timestamp)) BETWEEN %(from_date)s AND %(to_date)s
		{extra_sql}
		GROUP BY COALESCE(e.department, 'Unassigned'), {user_expr}, COALESCE(fv.me_activity_status, 'Unspecified')
		""",
		{
			"from_date": from_date,
			"to_date": to_date,
			**extra_params,
		},
		as_dict=True,
	)

	def apply_metrics(bucket, row):
		bucket.setdefault("active", 0)
		bucket.setdefault("inactive", 0)
		status = row.get("status")
		cnt = row.get("cnt", 0)
		if status == "Active":
			bucket["active"] += cnt
		elif status == "Inactive":
			bucket["inactive"] += cnt

	result_rows = _aggregate_section_user_rows(rows, apply_metrics)
	totals = {"active": 0, "inactive": 0, "total": 0}
	for row in result_rows:
		row["total"] = row.get("active", 0) + row.get("inactive", 0)
		totals["active"] += row.get("active", 0)
		totals["inactive"] += row.get("inactive", 0)
		totals["total"] += row.get("total", 0)

	return {"rows": result_rows, "totals": totals}


def _get_training_sessions(from_date, to_date, user=None, section=None):
	user_expr = "COALESCE(NULLIF(TRIM(fv.training_entry_filled_by), ''), fv.owner)"
	extra_conditions, extra_params = _section_user_filters(user=user, section=section)
	extra_sql = ""
	if extra_conditions:
		extra_sql = " AND " + " AND ".join(c.replace("{user_expr}", user_expr) for c in extra_conditions)

	rows = frappe.db.sql(
		f"""
		SELECT
			COALESCE(e.department, 'Unassigned') AS section,
			{user_expr} AS user,
			COALESCE(SUM(fv.training_no_of_schools_attended), 0) AS schools,
			COALESCE(SUM(fv.training_no_of_participants), 0) AS participants
		FROM `tabField Visit` fv
		{_employee_join(user_expr)}
		WHERE fv.docstatus < 2
		AND fv.type = 'Training'
		AND COALESCE(fv.training_date, DATE(fv.training_timestamp)) BETWEEN %(from_date)s AND %(to_date)s
		{extra_sql}
		GROUP BY COALESCE(e.department, 'Unassigned'), {user_expr}
		""",
		{
			"from_date": from_date,
			"to_date": to_date,
			**extra_params,
		},
		as_dict=True,
	)

	def apply_metrics(bucket, row):
		bucket["schools"] = bucket.get("schools", 0) + row.get("schools", 0)
		bucket["participants"] = bucket.get("participants", 0) + row.get("participants", 0)

	result_rows = _aggregate_section_user_rows(rows, apply_metrics)
	totals = {"schools": 0, "participants": 0}
	for row in result_rows:
		totals["schools"] += row.get("schools", 0)
		totals["participants"] += row.get("participants", 0)

	return {"rows": result_rows, "totals": totals}

File: tif_customization/tif_customization/page/field_staff_visit_mo/field_staff_visit_mo.js
frappe.pages['field-staff-visit-mo'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'SME Visits Summary',
		single_column: true
	});

	function month_start() {
		const dt = new Date();
		dt.setDate(1);
		return dt.toISOString().slice(0, 10);
	}

	const container = $(`
		<div class="field-staff-visit-report">
			<div class="filter-section" style="background: #f8f9fa; padding: 15px; border-radius: 6px; margin-bottom: 20px;">
				<div class="row">
					<div class="col-md-2">
						<label>${__('Visit From Date')}</label>
						<input type="date" id="fs-from-date" class="form-control" />
					</div>
					<div class="col-md-2">
						<label>${__('Visit To Date')}</label>
						<input type="date" id="fs-to-date" class="form-control" />
					</div>
					<div class="col-md-3">
						<label>${__('Section')}</label>
						<select id="fs-section" class="form-control">
							<option value="">${__('All Sections')}</option>
						</select>
					</div>
					<div class="col-md-3">
						<label>${__('User')}</label>
						<select id="fs-user" class="form-control">
							<option value="">${__('All Users')}</option>
						</select>
					</div>
					<div class="col-md-2" style="margin-top: 24px;">
						<button class="btn btn-primary" id="fs-apply">${__('Apply')}</button>
						<button class="btn btn-secondary" id="fs-reset">${__('Reset')}</button>
						<button class="btn btn-info" id="fs-print"><i class="fa fa-print"></i> ${__('Print')}</button>
					</div>
				</div>
			</div>
			<div id="fs-report"></div>
		</div>
	`);

	page.main.append(container);

	function build_section_select(selectedSection) {
		frappe.call({
			method: 'frappe.client.get_list',
			args: {
				doctype: 'Department',
				fields: ['name'],
				limit_page_length: 500,
				order_by: 'name asc'
			},
			callback: function(r) {
				const sections = r.message || [];
				const $section = $('#fs-section');
				$section.empty();
				$section.append(`<option value="">${__('All Sections')}</option>`);
				sections.forEach((s) => {
					const option = $(`<option value="${frappe.utils.escape_html(s.name)}">${frappe.utils.escape_html(s.name)}</option>`);
					if (s.name === selectedSection) {
						option.attr('selected', 'selected');
					}
					$section.append(option);
				});
			}
		});
	}

	function build_user_select(selectedUser) {
		frappe.call({
			method: 'frappe.client.get_list',
			args: {
				doctype: 'User',
				fields: ['name', 'full_name'],
				filters: { enabled: 1 },
				limit_page_length: 500,
				order_by: 'full_name asc'
			},
			callback: function(r) {
				const users = r.message || [];
				const $user = $('#fs-user');
				$user.empty();
				$user.append(`<option value="">${__('All Users')}</option>`);
				users.forEach(u => {
					const label = u.full_name ? `${u.full_name} (${u.name})` : u.name;
					const option = $(`<option value="${frappe.utils.escape_html(u.name)}">${frappe.utils.escape_html(label)}</option>`);
					if (u.name === selectedUser) {
						option.attr('selected', 'selected');
					}
					$user.append(option);
				});
			}
		});
	}

	function get_filters() {
		return {
			from_date: $('#fs-from-date').val(),
			to_date: $('#fs-to-date').val(),
			section: $('#fs-section').val() || '',
			user: $('#fs-user').val() || ''
		};
	}

	function load_report() {
		const filters = get_filters();
		if (!filters.from_date || !filters.to_date) {
			frappe.msgprint(__('Please select Visit From Date and Visit To Date.'));
			return;
		}

		$('#fs-report').html(`<p class="text-muted">${__('Loading report...')}</p>`);
		frappe.call({
			method: 'tif_customization.tif_customization.page.field_staff_visit_mo.field_staff_visit_mo.get_report_data',
			args: { filters },
			callback: function(r) {
				if (!r.message || r.message.error) {
					$('#fs-report').html(`<p class="text-danger">${__('Failed to load report.')}</p>`);
					return;
				}
				render_report(r.message);
			},
			error: function() {
				$('#fs-report').html(`<p class="text-danger">${__('Failed to load report.')}</p>`);
			}
		});
	}

	/*
	function add_submission_ratio(data, filters, callback) {
		const submittedUsers = get_submitted_users(data);

		frappe.call({
			method: 'frappe.client.get_list',
			args: {
				doctype: 'Has Role',
				fields: ['parent'],
				filters: {
					role: 'Field Staff',
					parenttype: 'User'
				},
				limit_page_length: 1000
			},
			callback: function(r) {
				const roleUsers = [...new Set((r.message || []).map(row => row.parent).filter(Boolean))];
				filter_enabled_field_staff_users(roleUsers, filters, function(eligibleUsers) {
					const eligibleCount = eligibleUsers.size;
					const submittedCount = submittedUsers.size;
					data.submission_ratio = {
						submitted_users: submittedCount,
						eligible_users: eligibleCount,
						ratio: eligibleCount ? Number((submittedCount / eligibleCount * 100).toFixed(1)) : 0
					};
					callback(data);
				});
			},
			error: function() {
				const submittedCount = submittedUsers.size;
				data.submission_ratio = {
					submitted_users: submittedCount,
					eligible_users: submittedCount,
					ratio: submittedCount ? 100 : 0
				};
				callback(data);
			}
		});
	}
	*/

	function filter_enabled_field_staff_users(roleUsers, filters, callback) {
		if (!roleUsers.length) {
			callback(new Set());
			return;
		}

		frappe.call({
			method: 'frappe.client.get_list',
			args: {
				doctype: 'User',
				fields: ['name'],
				filters: {
					name: ['in', roleUsers],
					enabled: 1,
					...(filters.user ? { name: filters.user } : {})
				},
				limit_page_length: 1000
			},
			callback: function(r) {
				const enabledUsers = new Set((r.message || []).map(row => row.name).filter(Boolean));
				if (!filters.section || !enabledUsers.size) {
					callback(enabledUsers);
					return;
				}

				frappe.call({
					method: 'frappe.client.get_list',
					args: {
						doctype: 'Employee',
						fields: ['user_id'],
						filters: {
							status: 'Active',
							department: filters.section,
							user_id: ['in', [...enabledUsers]]
						},
						limit_page_length: 1000
					},
					callback: function(employeeResponse) {
						callback(new Set((employeeResponse.message || []).map(row => row.user_id).filter(Boolean)));
					},
					error: function() {
						callback(enabledUsers);
					}
				});
			},
			error: function() {
				callback(new Set(roleUsers));
			}
		});
	}

	function get_submitted_users(data) {
		const submittedUsers = new Set();
		[data.marketing, data.me, data.training].forEach((group) => {
			(group?.rows || []).forEach((row) => {
				const user = String(row.user_name || row.user || '').trim();
				if (user && user !== 'Unassigned') {
					submittedUsers.add(user.toLowerCase());
				}
			});
		});
		return submittedUsers;
	}

	function rollup_marketing_by_section(rows) {
		const buckets = {};
		rows.forEach((row) => {
			const section = row.section || __('Unassigned');
			const bucket = buckets[section] || { section, new: 0, followup: 0, tps: 0, total: 0 };
			bucket.new += row.new || 0;
			bucket.followup += row.followup || 0;
			bucket.tps += row.tps || 0;
			bucket.total += row.total || 0;
			buckets[section] = bucket;
		});
		return Object.values(buckets).sort((a, b) => a.section.localeCompare(b.section));
	}

	function rollup_me_by_section(rows) {
		const buckets = {};
		rows.forEach((row) => {
			const section = row.section || __('Unassigned');
			const bucket = buckets[section] || { section, active: 0, inactive: 0, total: 0 };
			bucket.active += row.active || 0;
			bucket.inactive += row.inactive || 0;
			bucket.total += row.total || 0;
			buckets[section] = bucket;
		});
		return Object.values(buckets).sort((a, b) => a.section.localeCompare(b.section));
	}

	function rollup_training_by_section(rows) {
		const buckets = {};
		rows.forEach((row) => {
			const section = row.section || __('Unassigned');
			const bucket = buckets[section] || { section, schools: 0, participants: 0 };
			bucket.schools += row.schools || 0;
			bucket.participants += row.participants || 0;
			buckets[section] = bucket;
		});
		return Object.values(buckets).sort((a, b) => a.section.localeCompare(b.section));
	}

	function render_report(data) {
		const fromLabel = frappe.datetime.str_to_user(data.from_date || '');
		const toLabel = frappe.datetime.str_to_user(data.to_date || '');
		const header = `
			<div style="text-align: center; font-weight: bold; font-size: 20px; margin-bottom: 15px;">
				${__('SME Visits Summary')} (${fromLabel} ${__('to')} ${toLabel})
			</div>
		`;
		const submissionRatio = render_submission_ratio(data.submission_ratio);

		const marketingSection = render_marketing_section_table(data.marketing);
		const meSection = render_me_section_table(data.me);
		const trainingSection = render_training_section_table(data.training);

		const marketingUsers = render_marketing_user_table(data.marketing);
		// const meUsers = render_me_user_table(data.me);
		// const trainingUsers = render_training_user_table(data.training);

		$('#fs-report').html(`
			${header}
			${submissionRatio}
			<div class="row">
				<div class="col-md-4">${marketingSection}</div>
				<div class="col-md-4">${meSection}</div>
				<div class="col-md-4">${trainingSection}</div>
			</div>
			<div style="margin-top: 25px; margin-bottom: 10px; font-weight: bold; font-size: 16px;">
				${__('User Wise Detail')}
			</div>
			<div class="row">
				<div class="col-md-12 mb-3">${marketingUsers}</div>
				<!-- M&E Visits — User Wise table intentionally hidden -->
				<!-- Training Sessions — User Wise table intentionally hidden -->
			</div>
		`);
	}

	function render_submission_ratio(summary) {
		const submitted = summary?.submitted_users || 0;
		const eligible = summary?.eligible_users || 0;
		const ratio = Number(summary?.ratio || 0).toFixed(1);

		return `
			<div class="row" style="margin-bottom: 18px;">
				<div class="col-md-4">
					<div class="border rounded p-3 text-center">
						<div class="text-muted">${__('Submitted Users')}</div>
						<div style="font-size: 24px; font-weight: 700;">${submitted}</div>
					</div>
				</div>
				<div class="col-md-4">
					<div class="border rounded p-3 text-center">
						<div class="text-muted">${__('Eligible Users')}</div>
						<div style="font-size: 24px; font-weight: 700;">${eligible}</div>
					</div>
				</div>
				<div class="col-md-4">
					<div class="border rounded p-3 text-center">
						<div class="text-muted">${__('Submission Ratio')}</div>
						<div style="font-size: 24px; font-weight: 700;">${ratio}%</div>
					</div>
				</div>
			</div>
		`;
	}

	function render_marketing_section_table(marketing) {
		const rows = rollup_marketing_by_section(marketing?.rows || []);
		const totals = marketing?.totals || { new: 0, followup: 0, tps: 0, total: 0 };
		const body = rows.length ? rows.map(row => `
			<tr>
				<td>${frappe.utils.escape_html(row.section || '-')}</td>
				<td class="text-right">${row.new || 0}</td>
				<td class="text-right">${row.followup || 0}</td>
				<td class="text-right">${row.tps || 0}</td>
				<td class="text-right">${row.total || 0}</td>
			</tr>
		`).join('') : `<tr><td colspan="5" class="text-center">${__('No data')}</td></tr>`;

		return `
			<div class="table-responsive">
				<table class="table table-bordered table-striped" style="font-size: 12px;">
					<thead>
						<tr><th colspan="5" class="text-center">${__('Marketing Visits')}</th></tr>
						<tr>
							<th>${__('Section')}</th>
							<th>${__('New')}</th>
							<th>${__('Followup & Other Visits')}</th>
							<th>${__('TPS Visits')}</th>
							<th>${__('Grand Total')}</th>
						</tr>
					</thead>
					<tbody>${body}</tbody>
					<tfoot>
						<tr>
							<th>${__('Grand Total')}</th>
							<th class="text-right">${totals.new || 0}</th>
							<th class="text-right">${totals.followup || 0}</th>
							<th class="text-right">${totals.tps || 0}</th>
							<th class="text-right">${totals.total || 0}</th>
						</tr>
					</tfoot>
				</table>
			</div>
		`;
	}

	function render_me_section_table(me) {
		const rows = rollup_me_by_section(me?.rows || []);
		const totals = me?.totals || { active: 0, inactive: 0, total: 0 };
		const body = rows.length ? rows.map(row => `
			<tr>
				<td>${frappe.utils.escape_html(row.section || '-')}</td>
				<td class="text-right">${row.active || 0}</td>
				<td class="text-right">${row.inactive || 0}</td>
				<td class="text-right">${row.total || 0}</td>
			</tr>
		`).join('') : `<tr><td colspan="4" class="text-center">${__('No data')}</td></tr>`;

		return `
			<div class="table-responsive">
				<table class="table table-bordered table-striped" style="font-size: 12px;">
					<thead>
						<tr><th colspan="4" class="text-center">${__('M&E Visits')}</th></tr>
						<tr>
							<th>${__('Section')}</th>
							<th>${__('Active')}</th>
							<th>${__('Inactive')}</th>
							<th>${__('Grand Total')}</th>
						</tr>
					</thead>
					<tbody>${body}</tbody>
					<tfoot>
						<tr>
							<th>${__('Grand Total')}</th>
							<th class="text-right">${totals.active || 0}</th>
							<th class="text-right">${totals.inactive || 0}</th>
							<th class="text-right">${totals.total || 0}</th>
						</tr>
					</tfoot>
				</table>
			</div>
		`;
	}

	function render_training_section_table(training) {
		const rows = rollup_training_by_section(training?.rows || []);
		const totals = training?.totals || { schools: 0, participants: 0 };
		const body = rows.length ? rows.map(row => `
			<tr>
				<td>${frappe.utils.escape_html(row.section || '-')}</td>
				<td class="text-right">${row.schools || 0}</td>
				<td class="text-right">${row.participants || 0}</td>
			</tr>
		`).join('') : `<tr><td colspan="3" class="text-center">${__('No data')}</td></tr>`;

		return `
			<div class="table-responsive">
				<table class="table table-bordered table-striped" style="font-size: 12px;">
					<thead>
						<tr><th colspan="3" class="text-center">${__('Training Sessions')}</th></tr>
						<tr>
							<th>${__('Section')}</th>
							<th>${__('No. of Schools')}</th>
							<th>${__('No. of participants')}</th>
						</tr>
					</thead>
					<tbody>${body}</tbody>
					<tfoot>
						<tr>
							<th>${__('Grand Total')}</th>
							<th class="text-right">${totals.schools || 0}</th>
							<th class="text-right">${totals.participants || 0}</th>
						</tr>
					</tfoot>
				</table>
			</div>
		`;
	}

	function render_row_label(row) {
		const section = frappe.utils.escape_html(row.section || '-');
		const user = frappe.utils.escape_html(row.user_name || row.user || '-');
		return `<td>${section}</td><td>${user}</td>`;
	}

	function render_marketing_user_table(marketing) {
		const rows = marketing?.rows || [];
		const totals = marketing?.totals || { new: 0, followup: 0, tps: 0, total: 0 };
		const body = rows.length ? rows.map(row => `
			<tr>
				${render_row_label(row)}
				<td class="text-right">${row.new || 0}</td>
				<td class="text-right">${row.followup || 0}</td>
				<td class="text-right">${row.tps || 0}</td>
				<td class="text-right">${row.total || 0}</td>
			</tr>
		`).join('') : `<tr><td colspan="6" class="text-center">${__('No data')}</td></tr>`;

		return `
			<div class="table-responsive">
				<table class="table table-bordered table-striped" style="font-size: 12px;">
					<thead>
						<tr><th colspan="6" class="text-center">${__('Marketing Visits — User Wise')}</th></tr>
						<tr>
							<th>${__('Section')}</th>
							<th>${__('User')}</th>
							<th>${__('New')}</th>
							<th>${__('Followup & Other Visits')}</th>
							<th>${__('TPS Visits')}</th>
							<th>${__('Grand Total')}</th>
						</tr>
					</thead>
					<tbody>${body}</tbody>
					<tfoot>
						<tr>
							<th colspan="2">${__('Grand Total')}</th>
							<th class="text-right">${totals.new || 0}</th>
							<th class="text-right">${totals.followup || 0}</th>
							<th class="text-right">${totals.tps || 0}</th>
							<th class="text-right">${totals.total || 0}</th>
						</tr>
					</tfoot>
				</table>
			</div>
		`;
	}

	function render_me_user_table(me) {
		const rows = me?.rows || [];
		const totals = me?.totals || { active: 0, inactive: 0, total: 0 };
		const body = rows.length ? rows.map(row => `
			<tr>
				${render_row_label(row)}
				<td class="text-right">${row.active || 0}</td>
				<td class="text-right">${row.inactive || 0}</td>
				<td class="text-right">${row.total || 0}</td>
			</tr>
		`).join('') : `<tr><td colspan="5" class="text-center">${__('No data')}</td></tr>`;

		return `
			<div class="table-responsive">
				<table class="table table-bordered table-striped" style="font-size: 12px;">
					<thead>
						<tr><th colspan="5" class="text-center">${__('M&E Visits — User Wise')}</th></tr>
						<tr>
							<th>${__('Section')}</th>
							<th>${__('User')}</th>
							<th>${__('Active')}</th>
							<th>${__('Inactive')}</th>
							<th>${__('Grand Total')}</th>
						</tr>
					</thead>
					<tbody>${body}</tbody>
					<tfoot>
						<tr>
							<th colspan="2">${__('Grand Total')}</th>
							<th class="text-right">${totals.active || 0}</th>
							<th class="text-right">${totals.inactive || 0}</th>
							<th class="text-right">${totals.total || 0}</th>
						</tr>
					</tfoot>
				</table>
			</div>
		`;
	}

	function render_training_user_table(training) {
		const rows = training?.rows || [];
		const totals = training?.totals || { schools: 0, participants: 0 };
		const body = rows.length ? rows.map(row => `
			<tr>
				${render_row_label(row)}
				<td class="text-right">${row.schools || 0}</td>
				<td class="text-right">${row.participants || 0}</td>
			</tr>
		`).join('') : `<tr><td colspan="4" class="text-center">${__('No data')}</td></tr>`;

		return `
			<div class="table-responsive">
				<table class="table table-bordered table-striped" style="font-size: 12px;">
					<thead>
						<tr><th colspan="4" class="text-center">${__('Training Sessions — User Wise')}</th></tr>
						<tr>
							<th>${__('Section')}</th>
							<th>${__('User')}</th>
							<th>${__('No. of Schools')}</th>
							<th>${__('No. of participants')}</th>
						</tr>
					</thead>
					<tbody>${body}</tbody>
					<tfoot>
						<tr>
							<th colspan="2">${__('Grand Total')}</th>
							<th class="text-right">${totals.schools || 0}</th>
							<th class="text-right">${totals.participants || 0}</th>
						</tr>
					</tfoot>
				</table>
			</div>
		`;
	}

	function reset_filters() {
		$('#fs-from-date').val(month_start());
		$('#fs-to-date').val(frappe.datetime.get_today());
		$('#fs-section').val('');
		$('#fs-user').val('');
		load_report();
	}

	$('#fs-apply').on('click', load_report);
	$('#fs-reset').on('click', reset_filters);
	$('#fs-print').on('click', function() {
		window.print();
	});

	build_section_select('');
	build_user_select('');
	reset_filters();
};
